### PR TITLE
Refactor command runner interface, allow stdin writes

### PR DESCRIPTION
**Pull request by medyagh**
Initially this is preparing minikube to use kic ! (which requires setting StdIn on the commands)
previously minikube commands were all string, now it would be a exec.Cmd interface type.

This PR:

- add a new method to command runner to accept exec.Cmd as opposed to `cmd string`
- New Interface for Runner that returns a RunResult instead of err
```
type RunResult struct {
	Stdout   bytes.Buffer
	Stderr   bytes.Buffer
	ExitCode int
	Args     []string // the args that was passed to Runner
}

```
-  refactor and unexport funcs suchs as teeSSH.

### CODE REVIEW

File: pkg/minikube/command/command_runner.go [hunk 1]
```diff
@@ -17,15 +17,29 @@ limitations under the License.
...
 	"fmt"
 	"io"
+	"os/exec"
 	"path"
+	"strings"
 
 	"k8s.io/minikube/pkg/minikube/assets"
 )
 
+type RunResult struct {
```
Comment (josedonizetti) [inline pkg/minikube/command/command_runner.go]: missing struct and fields documentation

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go [hunk 2]
```diff
@@ -526,9 +539,10 @@ func (k *Bootstrapper) PullImages(k8s config.KubernetesConfig) error {
 		return fmt.Errorf("pull command is not supported by kubeadm v%s", version)
 	}
 
-	cmd := fmt.Sprintf("%s config images pull --config %s", invokeKubeadm(k8s.KubernetesVersion), yamlConfigPath)
-	if err := k.c.Run(cmd); err != nil {
-		return errors.Wrapf(err, "running cmd: %s", cmd)
+	cmd := command.ExecCmd(fmt.Sprintf("%s config images pull --config %s", invokeKubeadm(k8s.KubernetesVersion), yamlConfigPath))
+	rr, err := k.c.RunCmd(cmd)
+	if err != nil {
```
Comment (josedonizetti) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]:
```suggestion
	if rr, err := k.c.RunCmd(cmd); err != nil {
```

there are a few places I feel that writing like ^^ will improve readability, but this is nitpick

**Comment by medyagh** [comment]
you could add a cheap test here 

https://github.com/kubernetes/minikube/blob/master/test/integration/functional_test.go#L299
just basicly do `porfile list` and then make sure the output is a valid json.

File: .travis.yml [hunk 3]
```diff
@@ -36,6 +36,9 @@ matrix:
       script: make test
 after_success:
   - bash <(curl -s https://codecov.io/bash)
+travisBuddy:
+    regex: (FAIL:|\.go:\d+:|^panic:|failed$)
```
Comment (tstromberg) [inline .travis.yml]: Remove unrelated change

File: cmd/minikube/cmd/start.go [hunk 4]
```diff
@@ -1012,7 +1012,7 @@ func setupKubeAdm(mAPI libmachine.API, kc cfg.KubernetesConfig) bootstrapper.Boo
 	}
 	// Loads cached images, generates config files, download binaries
 	if err := bs.UpdateCluster(kc); err != nil {
-		exit.WithError("Failed to update cluster", err)
+		exit.WithError("Failed to update cluster<--Temproary logging-->", err)
```
Comment (tstromberg) [inline cmd/minikube/cmd/start.go]: Remove temp statement.

File: hack/jenkins/common.sh [hunk 5]
```diff
@@ -235,6 +235,9 @@ fi
 
 echo ""
 echo ">> Starting ${E2E_BIN} at $(date)"
+echo ">> The minikube version & commit id:"
```
Comment (tstromberg) [inline hack/jenkins/common.sh]: Remove unrelated change

File: pkg/drivers/none/none.go [hunk 6]
```diff
@@ -220,19 +222,19 @@ func (d *Driver) RunSSHCommandFromDriver() error {
...
-func stopKubelet(exec command.Runner) error {
+func stopKubelet(cr command.Runner) error {
 	glog.Infof("stopping kubelet.service ...")
 	stop := func() error {
-		cmdStop := "sudo systemctl stop kubelet.service"
-		cmdCheck := "sudo systemctl show -p SubState kubelet"
-		err := exec.Run(cmdStop)
-		if err != nil {
-			glog.Errorf("temporary error for %q : %v", cmdStop, err)
+		cmdStop := exec.Command("/bin/bash", "-c", "sudo systemctl stop kubelet.service")
```
Comment (tstromberg) [inline pkg/drivers/none/none.go]: Remove /bin/bash here and elsewhere - it isn't required unless there is shell logic to execute (`|, && or ||`) or variable substitution (`$`)

File: pkg/minikube/bootstrapper/certs.go [hunk 7]
```diff
@@ -318,21 +319,22 @@ func collectCACerts() (map[string]string, error) {
...
 	return stringHash, nil
 }
 
 // configureCACerts looks up and installs all uploaded PEM certificates in /usr/share/ca-certificates to system-wide certificate store (/etc/ssl/certs).
 // OpenSSL binary required in minikube ISO
-func configureCACerts(cmd command.Runner, caCerts map[string]string) error {
+func configureCACerts(cr command.Runner, caCerts map[string]string) error {
 	hasSSLBinary := true
-	if err := cmd.Run("which openssl"); err != nil {
+	c := exec.Command("/bin/bash", "-c", "which openssl")
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs.go]: optional: use "openssl version" instead.

File: pkg/minikube/bootstrapper/certs_test.go [hunk 8]
```diff
@@ -62,9 +62,9 @@ func TestSetupCerts(t *testing.T) {
 		certStorePath := path.Join(SSLCertStoreDir, dst)
 		certNameHash := "abcdef"
 		remoteCertHashLink := path.Join(SSLCertStoreDir, fmt.Sprintf("%s.0", certNameHash))
-		cmdMap[fmt.Sprintf("sudo ln -s '%s' '%s'", certFile, certStorePath)] = "1"
-		cmdMap[fmt.Sprintf("openssl x509 -hash -noout -in '%s'", certFile)] = certNameHash
-		cmdMap[fmt.Sprintf("sudo ln -s '%s' '%s'", certStorePath, remoteCertHashLink)] = "1"
+		cmdMap[fmt.Sprintf("/bin/bash -c sudo ln -s '%s' '%s'", certFile, certStorePath)] = "1"
+		cmdMap[fmt.Sprintf("/bin/bash -c openssl x509 -hash -noout -in '%s'", certFile)] = certNameHash
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/certs_test.go]: As mentioned elsewhere, none of these should be "/bin/bash -c". 

It's also worth mentioning that these commands appear to be incorrectly quoted.

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go [hunk 9]
```diff
@@ -136,12 +137,11 @@ func NewKubeadmBootstrapper(api libmachine.API) (*Bootstrapper, error) {
...
 func (k *Bootstrapper) GetKubeletStatus() (string, error) {
-	statusCmd := `sudo systemctl is-active kubelet`
-	status, err := k.c.CombinedOutput(statusCmd)
+	rr, err := k.c.RunCmd(exec.Command("/bin/bash", "-c", "sudo systemctl is-active kubelet"))
 	if err != nil {
-		return "", errors.Wrap(err, "getting status")
+		return "", errors.Wrapf(err, "getting kublet status. command: %q output: %q", rr.Command(), rr.Output())
 	}
-	s := strings.TrimSpace(status)
+	s := strings.TrimSpace(rr.Stdout.String() + rr.Stderr.String())
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: This should not parse stderr.

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go [hunk 10]
```diff
@@ -640,8 +642,8 @@ func (k *Bootstrapper) UpdateCluster(cfg config.KubernetesConfig) error {
 		}
 	}
 
-	if err := k.c.Run(`sudo systemctl daemon-reload && sudo systemctl start kubelet`); err != nil {
-		return errors.Wrap(err, "starting kubelet")
+	if rr, err := k.c.RunCmd(exec.Command("/bin/bash", "-c", "set -x;sudo systemctl daemon-reload && sudo systemctl start kubelet")); err != nil {
+		return errors.Wrapf(err, "starting kubelet, output: %s", rr.Output())
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: stderr is likely more important for error messages. Perhaps include by default?

File: pkg/minikube/cluster/mount.go [hunk 11]
```diff
@@ -131,20 +137,14 @@ func mntCmd(source string, target string, c *MountConfig) string {
...
-	cmd := umountCmd(target)
-	glog.Infof("Will run: %s", cmd)
-	out, err := r.CombinedOutput(cmd)
-	glog.Infof("unmount force err=%v, out=%s", err, out)
-	if err != nil {
-		return errors.Wrap(err, out)
+	// grep because findmnt will also display the parent!
+	c := exec.Command("/bin/bash", "-c", fmt.Sprintf("[ \"x$(findmnt -T %s | grep %s)\" != \"x\" ] && sudo umount -f %s || echo ", target, target, target))
+	if rr, err := r.RunCmd(c); err != nil {
+		glog.Infof("unmount force err=%v, out=%s", err, rr.Output())
```
Comment (tstromberg) [inline pkg/minikube/cluster/mount.go]: Log errors or return them, but not both, as it will eventually result in duplicate error messages being logged.

File: pkg/minikube/command/ssh_runner.go [hunk 12]
```diff
@@ -112,43 +123,27 @@ func (s *SSHRunner) Run(cmd string) error {
...
-	defer sess.Close()
-
-	var combined singleWriter
-	err = teeSSH(sess, cmd, &combined, &combined)
-	out := combined.b.String()
-	if err != nil {
-		return out, err
-	}
-	return out, nil
+// converts a exec.Cmd to string to be used by ssh runner
```
Comment (tstromberg) [inline pkg/minikube/command/ssh_runner.go]: Now that the mystery is solved, inline this function.

File: pkg/minikube/cruntime/cruntime_test.go [hunk 13]
```diff
@@ -194,6 +259,9 @@ func (f *FakeRunner) crio(args []string, _ bool) (string, error) {
 	if args[0] == "--version" {
 		return "crio version 1.13.0", nil
 	}
+	if args[0] == "something" { // doing this to suppress lint "result 1 (error) is always nil"
```
Comment (tstromberg) [inline pkg/minikube/cruntime/cruntime_test.go]: Use https://github.com/golangci/golangci-lint#nolint

File: pkg/drivers/none/none.go (hunk 6 above)
Comment (medyagh) [reply pkg/drivers/none/none.go]: exec.Command expects (name and args) would the name be "sudo" ?

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go (hunk 9 above)
Comment (medyagh) [reply pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: I was thinking maybe if the status is none, it would go to stderr ?

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go (hunk 10 above)
Comment (medyagh) [reply pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: do you mean in the RunResults.err , wrap the err with stderr ?

**Comment by medyagh** [comment]
the none failures, makes me doubt that something is wrong with this PR and None !

**Comment by medyagh** [comment]
one thought would be, the console would output a new line:

🔥  Auto-Detected KVM as driver. (Vish Version.. )
🔥  Creating kvm2 VM (CPUs=2, Memory=2000MB, Disk=20000MB) ...

File: pkg/minikube/cruntime/cri.go [hunk 14]
```diff
@@ -383,7 +399,11 @@ image-endpoint: unix://{{.Socket}}
 	if err := t.Execute(&b, opts); err != nil {
 		return err
 	}
-	return cr.Run(fmt.Sprintf("sudo mkdir -p %s && printf %%s \"%s\" | sudo tee %s", path.Dir(cPath), b.String(), cPath))
+	c := exec.Command("/bin/bash", "-c", fmt.Sprintf("sudo mkdir -p %s && printf %%s \"%s\" | sudo tee %s", path.Dir(cPath), b.String(), cPath))
+	if _, err := cr.RunCmd(c); err != nil {
+		return errors.Wrap(err, "populateCRIConfig")
```
Comment (tstromberg) [inline pkg/minikube/cruntime/cri.go]: Like elsewhere, this code doesn't know it's own function name, so this will likely go stale and get repeated up the error chain. Just say "run" here, and let the caller of populateCRIConfig wrap it with the name it knows this code by.

File: pkg/minikube/cruntime/cruntime_test.go [hunk 15]
```diff
@@ -259,10 +323,15 @@ func (f *FakeRunner) systemctl(args []string, root bool) (string, error) {
 	out := ""
 
 	for i, arg := range args {
+		// shamelessly useless if statement, only to suppress the lint : - result 0 (string) is always ""
```
Comment (tstromberg) [inline pkg/minikube/cruntime/cruntime_test.go]: remove and use //nolint

File: pkg/minikube/cruntime/docker.go [hunk 16]
```diff
@@ -136,7 +157,11 @@ func (r *Docker) StopContainers(ids []string) error {
 		return nil
 	}
 	glog.Infof("Stopping containers: %s", ids)
-	return r.Runner.Run(fmt.Sprintf("docker stop %s", strings.Join(ids, " ")))
+	c := exec.Command("docker", "stop", shellquote.Join(ids...))
```
Comment (tstromberg) [inline pkg/minikube/cruntime/docker.go]: ids should remain separate arguments.

File: pkg/minikube/cruntime/docker.go [hunk 17]
```diff
@@ -127,7 +143,12 @@ func (r *Docker) KillContainers(ids []string) error {
 		return nil
 	}
 	glog.Infof("Killing containers: %s", ids)
-	return r.Runner.Run(fmt.Sprintf("docker rm -f %s", strings.Join(ids, " ")))
+	args := append([]string{"rm", "-f"}, ids...)
```
Comment (tstromberg) [inline pkg/minikube/cruntime/docker.go]: ids should remain separate arguments.

File: pkg/minikube/problem/err_map.go [hunk 18]
```diff
@@ -56,7 +56,7 @@ var vmProblems = map[string]match{
 		Issues: []int{1926, 4206},
 	},
 	"HYPERKIT_NOT_FOUND": {
-		Regexp:         re(`Driver "hyperkit" not found. Do you have the plugin binary .* accessible in your PATH?`),
```
Comment (tstromberg) [inline pkg/minikube/problem/err_map.go]: unrelated change?

File: pkg/drivers/none/none.go [hunk 19]
```diff
@@ -248,13 +250,21 @@ func stopKubelet(exec command.Runner) error {
 }
 
 // restartKubelet restarts the kubelet
-func restartKubelet(exec command.Runner) error {
+func restartKubelet(cr command.Runner) error {
 	glog.Infof("restarting kubelet.service ...")
-	return exec.Run("sudo systemctl restart kubelet.service")
+	c := exec.Command("sudo", "systemctl", "restart", "kubelet.service")
+	if _, err := cr.RunCmd(c); err != nil {
+		return errors.Wrapf(err, "restartKubelet")
```
[the remark this answers is not in the document]
Comment (medyagh) [reply pkg/drivers/none/none.go]: good point, I will just return err, since err includings stderr now.

File: pkg/minikube/cruntime/docker.go (hunk 17 above)
Comment (medyagh) [reply pkg/minikube/cruntime/docker.go]: @tstromberg I don't undrestand what you mean. do you mind explaining ?

File: pkg/minikube/problem/err_map.go (hunk 18 above)
Comment (medyagh) [reply pkg/minikube/problem/err_map.go]: yeah but the change actually is good.

Comment (medyagh) [reply pkg/minikube/problem/err_map.go]: that regex on master wont match the error

**Comment by medyagh** [comment]
- the test for kvm passes. ( with one addon flake which I think is a known flake)
- the test for hyperkit mac os passes on my laptop. but CI seems to be failing for everyone
- the none test passes except for "oldest supported version "

File: pkg/drivers/none/none.go [hunk 20]
```diff
@@ -217,19 +218,19 @@ func (d *Driver) RunSSHCommandFromDriver() error {
...
-		errStatus := exec.CombinedOutputTo(cmdCheck, &out)
-		if errStatus != nil {
-			glog.Errorf("temporary error: for %q : %v", cmdCheck, errStatus)
+		cmd = exec.Command("sudo", "systemctl", "show", "-p", "SubState", "kubelet")
+		cmd.Stdout = &out
+		cmd.Stderr = &out
+		if rr, err := cr.RunCmd(cmd); err != nil {
+			glog.Errorf("temporary error: for %q : %v", rr.Command(), err)
 		}
 		if !strings.Contains(out.String(), "dead") && !strings.Contains(out.String(), "failed") {
```
Comment (tstromberg) [inline pkg/drivers/none/none.go]: Can this use rr.Stdout now?

File: pkg/minikube/command/command_runner.go [hunk 21]
```diff
@@ -17,34 +17,28 @@ limitations under the License.
...
-	// Or, you can set out to os.Stdout, the command output and
-	// error would show on your terminal immediately before you
-	// cmd exit. This is useful for a long run command such as
-	// continuously print running logs.
-	CombinedOutputTo(cmd string, out io.Writer) error
-
-	// CombinedOutput runs the command and returns its combined standard
-	// output and standard error.
-	CombinedOutput(cmd string) (string, error)
+	// RunCmd is a new expermintal way to run commands, takes Cmd interface and returns run result.
```
Comment (tstromberg) [inline pkg/minikube/command/command_runner.go]: Please update this comment.

File: pkg/minikube/cruntime/containerd.go [hunk 22]
```diff
@@ -194,18 +205,30 @@ func (r *Containerd) Enable(disOthers bool) error {
 		return err
 	}
 	// Otherwise, containerd will fail API requests with 'Unimplemented'
-	return r.Runner.Run("sudo systemctl restart containerd")
+	c := exec.Command("sudo", "systemctl", "restart", "containerd")
+	if _, err := r.Runner.RunCmd(c); err != nil {
+		return errors.Wrap(err, "enable containrd.")
```
Comment (tstromberg) [inline pkg/minikube/cruntime/containerd.go]: Fix error message: this should say restart containerd.

File: pkg/minikube/cruntime/containerd.go [hunk 23]
```diff
@@ -194,18 +205,30 @@ func (r *Containerd) Enable(disOthers bool) error {
...
+	}
+	return nil
 }
 
 // Disable idempotently disables containerd on a host
 func (r *Containerd) Disable() error {
-	return r.Runner.Run("sudo systemctl stop containerd")
+	c := exec.Command("sudo", "systemctl", "stop", "containerd")
+	if _, err := r.Runner.RunCmd(c); err != nil {
+		return errors.Wrapf(err, "disable containrd.")
```
Comment (tstromberg) [inline pkg/minikube/cruntime/containerd.go]: this should say stop containerd. Also, no need for trailing period.

File: pkg/minikube/cruntime/containerd.go [hunk 24]
```diff
@@ -194,18 +205,30 @@ func (r *Containerd) Enable(disOthers bool) error {
...
+	return nil
 }
 
 // LoadImage loads an image into this runtime
 func (r *Containerd) LoadImage(path string) error {
 	glog.Infof("Loading image: %s", path)
-	return r.Runner.Run(fmt.Sprintf("sudo ctr -n=k8s.io images import %s", path))
+	c := exec.Command("ctr", "-n=k8s.io", "images", "import", path)
+	if _, err := r.Runner.RunCmd(c); err != nil {
+		return errors.Wrapf(err, "disable containrd.")
```
Comment (tstromberg) [inline pkg/minikube/cruntime/containerd.go]: this should say ctr images import

File: pkg/minikube/cruntime/crio.go [hunk 25]
```diff
@@ -70,12 +73,18 @@ func (r *CRIO) DefaultCNI() bool {
 
 // Available returns an error if it is not possible to use this runtime on a host
 func (r *CRIO) Available() error {
-	return r.Runner.Run("command -v crio")
+	c := exec.Command("command", "-v", "crio")
```
Comment (tstromberg) [inline pkg/minikube/cruntime/crio.go]: I believe this still needs bash: command is a shell built-in.

File: pkg/minikube/cruntime/crio.go [hunk 26]
```diff
@@ -95,18 +104,29 @@ func (r *CRIO) Enable(disOthers bool) error {
...
+	return nil
 }
 
 // LoadImage loads an image into this runtime
 func (r *CRIO) LoadImage(path string) error {
 	glog.Infof("Loading image: %s", path)
-	return r.Runner.Run(fmt.Sprintf("sudo podman load -i %s", path))
+	c := exec.Command("sudo", "podman", "load", "-i", path)
+	if _, err := r.Runner.RunCmd(c); err != nil {
+		return errors.Wrap(err, "LoadImage crio.")
```
Comment (tstromberg) [inline pkg/minikube/cruntime/crio.go]: remove trailing period in error messages (here and elsewhere).

File: pkg/minikube/command/exec_runner.go [hunk 27]
```diff
@@ -35,41 +37,45 @@ import (
...
+	} else {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			rr.ExitCode = exitError.ExitCode()
+		}
+		glog.Infof("(ExecRunner) Non-zero exit: %v: %v (%s)\n%s", rr.Command(), err, elapsed, rr.Output())
+		err = errors.Wrapf(err, fmt.Sprintf("stderr: %s", rr.Stderr.String()))
 	}
-	return b.String(), nil
-
+	return rr, err
```
Comment (tstromberg) [inline pkg/minikube/command/exec_runner.go]: Would it be possible to include the same context into the error message that the previous implementation had?

`return errors.Wrapf(err, "command failed: %s\nstdout: %s\nstderr: %s", cmd, outB.String(), errB.String())`

File: pkg/minikube/cruntime/containerd.go [hunk 28]
```diff
@@ -194,18 +205,30 @@ func (r *Containerd) Enable(disOthers bool) error {
...
+	}
+	return nil
 }
 
 // LoadImage loads an image into this runtime
 func (r *Containerd) LoadImage(path string) error {
 	glog.Infof("Loading image: %s", path)
-	return r.Runner.Run(fmt.Sprintf("sudo ctr -n=k8s.io images import %s", path))
+	c := exec.Command("ctr", "-n=k8s.io", "images", "import", path)
```
[the remark this answers is not in the document]
Comment (medyagh) [reply pkg/minikube/cruntime/containerd.go]: ah good catch !

File: pkg/minikube/cruntime/crio.go (hunk 25 above)
Comment (medyagh) [reply pkg/minikube/cruntime/crio.go]: how about using "which" ?

File: pkg/drivers/none/none.go (hunk 20 above)
Comment (medyagh) [reply pkg/drivers/none/none.go]: we can !

**Comment by tstromberg** [comment]
```
# k8s.io/minikube/pkg/minikube/command
pkg/minikube/command/exec_runner.go:75: Wrapf format %s has arg cmd of wrong type *os/exec.Cmd
```

File: pkg/drivers/none/none.go [hunk 29]
```diff
@@ -229,10 +229,11 @@ func stopKubelet(cr command.Runner) error {
 		cmd = exec.Command("sudo", "systemctl", "show", "-p", "SubState", "kubelet")
 		cmd.Stdout = &out
```
Comment (tstromberg) [inline pkg/drivers/none/none.go]: The `out` variable is no longer necessary.

**Comment by medyagh** [comment]
Weird error on Linux virtualbox integration test:

```
        --- FAIL: TestDownloadOnly/group/ExpectedDefaultDriver (0.05s)
            a_serial_test.go:97: (dbg) Run:  out/minikube-linux-amd64 profile list --output json
            a_serial_test.go:115: got driver "virtualbox", expected "kvm2"
                start output: -- stdout --
                * [download-20191029T234149.569978431-6061] minikube v1.5.0 on Ubuntu 18.04
                  - KUBECONFIG=/home/minikubeci/minikube-integration/linux-amd64-virtualbox-5530-3609-773c525ae71c1852ac650b72543c8a695a1431d1/kubeconfig
                  - MINIKUBE_BIN=out/minikube-linux-amd64
                  - MINIKUBE_HOME=/home/minikubeci/minikube-integration/linux-amd64-virtualbox-5530-3609-773c525ae71c1852ac650b72543c8a695a1431d1/.minikube
                  - MINIKUBE_LOCATION=5530
                * Automatically selected the 'virtualbox' driver (alternates: [none])
                * Downloading VM boot image ...
                * Downloading kubelet v1.11.10
                * Downloading kubeadm v1.11.10
                * Download complete!
                 
```

**Comment by tstromberg** [comment]
Flakes appear unrelated. Merging.

**Comment by tstromberg** [comment]
> Weird error on Linux virtualbox integration test:
> since it shows its running on `GPU linux` machine I suspect the GPU linux machine does not have KVM running ??

I believe I've fixed this. It seemed that the GPU machine had a bad KVM-SPICE config. I deinstalled the SPICE driver.
